### PR TITLE
Add bitcoin 22.0

### PR DESCRIPTION
**Pull request by nolim1t**
Add bitcoin 22.0 to the containers. Release notes [here](https://bitcoincore.org/en/releases/22.0/)

### CODE REVIEW

**Comment by nolim1t** [comment]
Seems theres no GPG signature available

```
#22 [preparer-release 1/5] ADD https://bitcoincore.org/bin/bitcoin-core-22.0/SHA256SUMS.asc  ./
#22 sha256:f378287679c7ef1c013b2cf7e073f7c5e3f9a7c0880ef5362372acee9d91bfe9
#22 DONE 0.0s

#24 [preparer-release 2/5] ADD https://bitcoincore.org/bin/bitcoin-core-22.0/bitcoin-22.0.tar.gz ./
#24 sha256:f61324e34a7a79e0a5abcd55e60d99a729191e8a5e0e4c3b89b2dde22fd8da01
#24 DONE 0.1s

#25 [preparer-release 3/5] RUN gpg --verify SHA256SUMS.asc
#25 sha256:6cf18865cb3be471e9b9217a0cf42d84dd1728a8ec169ef94cea14bd9ec8ea4e
#25 0.254 gpg: no signed data
#25 0.254 gpg: can't hash datafile: No data
#25 ERROR: executor failed running [/bin/sh -c gpg --verify SHA256SUMS.asc]: exit code: 2
```

**Comment by nolim1t** [comment]
Raised issue on the bitcoin core website https://github.com/bitcoin-core/bitcoincore.org/issues/798

**Comment by nolim1t** [comment]
Appears the verification file is split into two pieces..  awaiting feedback on how to verify something like this.

Seems ```gpg --verify``` fails

**Comment by meeDamian** [comment]
First, way more keys needs to be imported:

```Docker
ENV KEYS 71A3B16735405025D447E8F274810B012346C9A6 01EA5486DE18A882D4C2684590C8019E36C2E964 0CCBAAFD76A2ECE2CCD3141DE2FFD5B1D88CA97D 152812300785C96444D3334D17565732E08E5E41 0AD83877C1F0CD1EE9BD660AD7CC770B81FD22A8 590B7292695AFFA5B672CBB2E13FC145CD3F4304 28F5900B1BB5D1A4B6B6D1A9ED357015286A333D CFB16E21C950F67FA95E558F2EEB9F5CC09526C1 6E01EEC9656903B0542B8F1003DB6322267C373B D1DBF2C4B96F2DEBF4C16654410108112E7EA81F 9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C 74E2DEF5D77260B98BC19438099BAD163C70FBFA 
```

These 2 don't seem to be available on keyserver.ubuntu.com, so:
```Docker
ENV KEYS2 637DB1E23370F84AFF88CCE03152347D07DA627C 82921A4B88FD454B7EB8CE3C796C4109063D4EAF 
```

Then actually import them:

```Docker
RUN timeout 16s gpg --recv-keys $KEYS2
```

Need to also`ADD` this file:

```Docker
ADD https://bitcoincore.org/bin/bitcoin-core-$VERSION/SHA256SUMS ./
```

Finally, to verify:

```Docker
RUN gpg --verify SHA256SUMS.asc SHA256SUMS
```

Worked for me (even though @achow101 key seems to be expired):

```sh
$ gpg --verify SHA256SUMS.asc SHA256SUMS                                                                                            
gpg: Signature made Fri 10 Sep 2021 07:29:17 PM +08
gpg:                using RSA key 0CCBAAFD76A2ECE2CCD3141DE2FFD5B1D88CA97D
gpg: Good signature from ".0xB10C <0xb10c@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 982A 193E 3CE0 EED5 35E0  9023 188C BB26 4841 6AD5
     Subkey fingerprint: 0CCB AAFD 76A2 ECE2 CCD3  141D E2FF D5B1 D88C A97D
gpg: Signature made Fri 10 Sep 2021 04:09:04 AM +08
gpg:                using RSA key 152812300785C96444D3334D17565732E08E5E41
gpg:                issuer "achow101@gmail.com"
gpg: Good signature from "Andrew Chow (Official New Key) <achow101@gmail.com>" [expired]
gpg:                 aka "Andrew Chow <achow101-github@achow101.com>" [expired]
gpg:                 aka "Andrew Chow <achow101-lists@achow101.com>" [expired]
gpg:                 aka "Andrew Chow <andrew@achow101.com>" [expired]
gpg:                 aka "Andrew Chow <achow101@yahoo.com>" [expired]
gpg: Note: This key has expired!
Primary key fingerprint: 1528 1230 0785 C964 44D3  334D 1756 5732 E08E 5E41
gpg: Signature made Fri 10 Sep 2021 04:16:18 AM +08
gpg:                using RSA key 0AD83877C1F0CD1EE9BD660AD7CC770B81FD22A8
gpg:                issuer "benthecarman@live.com"
gpg: Good signature from "Ben Carman <benthecarman@live.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 0AD8 3877 C1F0 CD1E E9BD  660A D7CC 770B 81FD 22A8
gpg: Signature made Fri 10 Sep 2021 09:00:35 PM +08
gpg:                using RSA key 590B7292695AFFA5B672CBB2E13FC145CD3F4304
gpg:                issuer "darosior@protonmail.com"
gpg: Good signature from "Antoine Poinsot <darosior@protonmail.com>" [unknown]
gpg:                 aka "Antoine Poinsot <antoine@revault.dev>" [unknown]
gpg:                 aka "darosior <darosior@ln.dev>" [unknown]
gpg:                 aka "darosior <darosior@protonmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 590B 7292 695A FFA5 B672  CBB2 E13F C145 CD3F 4304
gpg: Signature made Fri 10 Sep 2021 04:54:01 AM +08
gpg:                using RSA key 28F5900B1BB5D1A4B6B6D1A9ED357015286A333D
gpg: Good signature from "Duncan Dean <duncangleeddean@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 9484 44FC E03B 05BA 5AB0  591E C37B 1C1D 44C7 86EE
     Subkey fingerprint: 28F5 900B 1BB5 D1A4 B6B6  D1A9 ED35 7015 286A 333D
gpg: Signature made Fri 10 Sep 2021 10:26:03 PM +08
gpg:                using RSA key 637DB1E23370F84AFF88CCE03152347D07DA627C
gpg: Good signature from "Stephan Oeste (it) <it@oeste.de>" [unknown]
gpg:                 aka "Emzy E. (emzy) <emzy@emzy.de>" [unknown]
gpg:                 aka "Stephan Oeste (Master-key) <stephan@oeste.de>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 9EDA FF80 E080 6596 04F4  A76B 2EBB 056F D847 F8A7
     Subkey fingerprint: 637D B1E2 3370 F84A FF88  CCE0 3152 347D 07DA 627C
gpg: Signature made Fri 10 Sep 2021 09:04:14 AM +08
gpg:                using RSA key CFB16E21C950F67FA95E558F2EEB9F5CC09526C1
gpg:                issuer "fanquake@gmail.com"
gpg: Good signature from "Michael Ford (bitcoin-otc) <fanquake@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: E777 299F C265 DD04 7930  70EB 944D 35F9 AC3D B76A
     Subkey fingerprint: CFB1 6E21 C950 F67F A95E  558F 2EEB 9F5C C095 26C1
gpg: Signature made Fri 10 Sep 2021 04:03:16 PM +08
gpg:                using RSA key 6E01EEC9656903B0542B8F1003DB6322267C373B
gpg:                issuer "gugger@gmail.com"
gpg: Good signature from "Oliver Gugger <gugger@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: F4FC 70F0 7310 0284 24EF  C20A 8E42 5659 3F17 7720
     Subkey fingerprint: 6E01 EEC9 6569 03B0 542B  8F10 03DB 6322 267C 373B
gpg: Signature made Fri 10 Sep 2021 04:07:53 AM +08
gpg:                using RSA key D1DBF2C4B96F2DEBF4C16654410108112E7EA81F
gpg:                issuer "hebasto@gmail.com"
gpg: Good signature from "Hennadii Stepanov (GitHub key) <32963518+hebasto@users.noreply.github.com>" [unknown]
gpg:                 aka "Hennadii Stepanov (hebasto) <hebasto@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: D1DB F2C4 B96F 2DEB F4C1  6654 4101 0811 2E7E A81F
gpg: Signature made Fri 10 Sep 2021 03:14:14 PM +08
gpg:                using RSA key 82921A4B88FD454B7EB8CE3C796C4109063D4EAF
gpg:                issuer "jon@atack.com"
gpg: Good signature from "Jon Atack <jon@atack.com>" [unknown]
gpg:                 aka "jonatack <jon@atack.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 8292 1A4B 88FD 454B 7EB8  CE3C 796C 4109 063D 4EAF
gpg: Signature made Sat 11 Sep 2021 01:33:30 AM +08
gpg:                using RSA key 9DEAE0DC7063249FB05474681E4AED62986CD25D
gpg: Good signature from "Wladimir J. van der Laan <laanwj@visucore.com>" [unknown]
gpg:                 aka "Wladimir J. van der Laan <laanwj@gmail.com>" [unknown]
gpg:                 aka "Wladimir J. van der Laan <laanwj@protonmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 71A3 B167 3540 5025 D447  E8F2 7481 0B01 2346 C9A6
     Subkey fingerprint: 9DEA E0DC 7063 249F B054  7468 1E4A ED62 986C D25D
gpg: Signature made Fri 10 Sep 2021 04:22:36 AM +08
gpg:                using RSA key 9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C
gpg:                issuer "aaron@sipsorcery.com"
gpg: Good signature from "Aaron Clauson (sipsorcery) <aaron@sipsorcery.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 9D3C C86A 72F8 4943 42EA  5FD1 0A41 BDC3 F4FA FF1C
gpg: Signature made Fri 10 Sep 2021 05:59:33 PM +08
gpg:                using RSA key 74E2DEF5D77260B98BC19438099BAD163C70FBFA
gpg:                issuer "will8clark@gmail.com"
gpg: Good signature from "Will Clark <will8clark@gmail.com>" [unknown]
gpg: WARNING: This key is not certified with a trusted signature!
gpg:          There is no indication that the signature belongs to the owner.
Primary key fingerprint: 74E2 DEF5 D772 60B9 8BC1  9438 099B AD16 3C70 FBFA

$ echo $?                                                                                                                              
0
```

**Comment by meeDamian** [comment]
> First, way more keys needs to be imported:
> 
> ```
> ENV KEYS 71A3B16735405025D447E8F274810B012346C9A6 01EA5486DE18A882D4C2684590C8019E36C2E964 0CCBAAFD76A2ECE2CCD3141DE2FFD5B1D88CA97D 152812300785C96444D3334D17565732E08E5E41 0AD83877C1F0CD1EE9BD660AD7CC770B81FD22A8 590B7292695AFFA5B672CBB2E13FC145CD3F4304 28F5900B1BB5D1A4B6B6D1A9ED357015286A333D CFB16E21C950F67FA95E558F2EEB9F5CC09526C1 6E01EEC9656903B0542B8F1003DB6322267C373B D1DBF2C4B96F2DEBF4C16654410108112E7EA81F 9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C 74E2DEF5D77260B98BC19438099BAD163C70FBFA 
> ```
> 
> These 2 don't seem to be available on keyserver.ubuntu.com, so:
> 
> ```
> ENV KEYS2 637DB1E23370F84AFF88CCE03152347D07DA627C 82921A4B88FD454B7EB8CE3C796C4109063D4EAF 
> ```

Would be nice if one could specify threshold, ex: at least 5 sig validations required.

**Comment by meeDamian** [comment]
And sorry it's not a PR, I'm on a 📱.

**Comment by nolim1t** [comment]
Added some more keys in

**Comment by Goro2030** [comment]
This "signatures" issue is highly irregular for a so-well-known piece of software coded by cryptographers ... I won't install the new version until this conundrum is sorted out.

**Comment by nolim1t** [comment]
> new 

Signature thing verified now.. just a different method. I don't know why they changed the way things are done. 

See the linked ticket

**Comment by nolim1t** [comment]
New verification process in case anyone else is interested

https://github.com/bitcoin/bitcoin/issues/22634

File: 22.0/Dockerfile [hunk 1]
```diff
@@ -0,0 +1,250 @@
...
+
+RUN apk add --no-cache  gnupg
+
+# Bitcoin keys
+ENV KEYS 71A3B16735405025D447E8F274810B012346C9A6 01EA5486DE18A882D4C2684590C8019E36C2E964 0CCBAAFD76A2ECE2CCD3141DE2FFD5B1D88CA97D 152812300785C96444D3334D17565732E08E5E41 0AD83877C1F0CD1EE9BD660AD7CC770B81FD22A8 590B7292695AFFA5B672CBB2E13FC145CD3F4304 28F5900B1BB5D1A4B6B6D1A9ED357015286A333D CFB16E21C950F67FA95E558F2EEB9F5CC09526C1 6E01EEC9656903B0542B8F1003DB6322267C373B D1DBF2C4B96F2DEBF4C16654410108112E7EA81F 9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C 74E2DEF5D77260B98BC19438099BAD163C70FBFA
+# New keys
+ENV KEYS2 637DB1E23370F84AFF88CCE03152347D07DA627C 82921A4B88FD454B7EB8CE3C796C4109063D4EAF
+
+RUN timeout 32s gpg  --keyserver keyserver.ubuntu.com  --recv-keys $KEYS
+RUN timeout 32s gpg --keyserver keys.openpgp.org --recv-keys $KEYS2
```
[the remark this answers is not in the document]
Comment (meeDamian) [reply 22.0/Dockerfile:45]: Also I'm not saying I have OCD, and the difference in spacing bothers me more than it should, but… 🙏

File: 22.0/Dockerfile [hunk 2]
```diff
@@ -0,0 +1,250 @@
...
+#
+FROM alpine:${VER_ALPINE} AS preparer-base
+
+# Make sure APKs are downloaded over SSL. See: https://github.com/gliderlabs/docker-alpine/issues/184
+RUN sed -i 's|http://dl-cdn.alpinelinux.org|https://alpine.global.ssl.fastly.net|g' /etc/apk/repositories
+
+RUN apk add --no-cache  gnupg
+
+# Bitcoin keys
+ENV KEYS 71A3B16735405025D447E8F274810B012346C9A6 01EA5486DE18A882D4C2684590C8019E36C2E964 0CCBAAFD76A2ECE2CCD3141DE2FFD5B1D88CA97D 152812300785C96444D3334D17565732E08E5E41 0AD83877C1F0CD1EE9BD660AD7CC770B81FD22A8 590B7292695AFFA5B672CBB2E13FC145CD3F4304 28F5900B1BB5D1A4B6B6D1A9ED357015286A333D CFB16E21C950F67FA95E558F2EEB9F5CC09526C1 6E01EEC9656903B0542B8F1003DB6322267C373B D1DBF2C4B96F2DEBF4C16654410108112E7EA81F 9D3CC86A72F8494342EA5FD10A41BDC3F4FAFF1C 74E2DEF5D77260B98BC19438099BAD163C70FBFA
```
[the remark this answers is not in the document]
Comment (nolim1t) [reply 22.0/Dockerfile:40]: Cool.. do you know a command (i.e. `sort`) which can do this ?

File: 22.0/Dockerfile (hunk 1 above)
[the remark this answers is not in the document]
Comment (nolim1t) [reply 22.0/Dockerfile:45]: Ok i'll check it out. Less is more

**Comment by nolim1t** [comment]
Looks like not all the keys are available on the keyserver

**Comment by ParkaMark** [comment]
I must say, this has been one of the most PITA changes I've had to deal with, with regards to automating the polling, verification and then download and installation of bitcoind - but the script below now achieves what I wanted.

Note I had to use some prime number version construction in order to make the version string "22.0" (which is how this new version has been released on the website) to equal the version string "22.0.0" which is how the version is reported when running "bitcoind -version" on the command line. The version function effectively resolves both of these strings to the same unique number where I can then test for equality.

They gpg keys issue still looks very flakey - I managed to import all the necessary keys manually outside of this script from various sources/posts located elsewhere, but the verification now works as expected.

So posting the below incase anyone finds this useful.

```
#!/bin/bash

version() {
  cat | awk -F . '{print 2^$1 * 3^$2 * 5^$3 * 7^$4}'
}

latest=$(curl -s https://bitcoincore.org/en/download/ | egrep -o 'Latest version: [0-9.]+' | awk '{print $NF}')
[[ -z $latest ]] && exit 1
[[ $(~/bin/bitcoind -version | head -1 | egrep -o '[0-9.]+' | version) -eq $(echo $latest | version) ]] && exit 0

for x in SHA256SUMS{,.asc}
do
  wget -qN https://bitcoincore.org/bin/bitcoin-core-${latest}/${x}
done

gpg --verify SHA256SUMS.asc SHA256SUMS && \
wget -q https://bitcoincore.org/bin/bitcoin-core-${latest}/bitcoin-${latest}-x86_64-linux-gnu.tar.gz && \
sha256sum --ignore-missing --check SHA256SUMS && \
tar zxvf bitcoin-${latest}-x86_64-linux-gnu.tar.gz && \
rm -vrf bin && \
mv -v bitcoin-*/bin . && \
rm -vrf bitcoin-* && \
/script/to/restart/bitcoind
```

**Review by Jess25051980 — APPROVED** [review]

**Pull request merged**